File: apps/native/android/app/src/main/java/raven/thecommit/company/RavenDownloadPlugin.java
package raven.thecommit.company;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/** File download with progress and cancel; the Filesystem plugin's downloadFile has neither. */
@CapacitorPlugin(name = "RavenDownload")
public class RavenDownloadPlugin extends Plugin {
    // Live connections by download id; cancel disconnects, which fails the read at once even when stalled.
    private final Map<String, HttpURLConnection> connections = new ConcurrentHashMap<>();

    // download({ id, url, name, headers }) writes cache/media/<folder>/<file> and resolves { path } when complete.
    // The page names the file but never the location; only http(s) sources are fetched.
    @PluginMethod
    public void download(PluginCall call) {
        String id = call.getString("id", "");
        String name = call.getString("name", "");
        URL url;
        try { url = new URL(call.getString("url", "")); } catch (Exception e) { call.reject("bad url"); return; }
        // name is <folder>/<file>: one level, no traversal.
        if (!name.matches("[^/\\\\.][^/\\\\]*/[^/\\\\.][^/\\\\]*") || !url.getProtocol().matches("https?")) {
            call.reject("bad name or url");
            return;
        }
        File target = new File(new File(getContext().getCacheDir(), "media"), name);
        File part = new File(target.getPath() + ".part");
        new Thread(() -> {
            HttpURLConnection connection = null;
            try {
                connection = (HttpURLConnection) url.openConnection();
                connection.setConnectTimeout(15_000);
                connection.setReadTimeout(30_000);
                connections.put(id, connection);
                JSObject headers = call.getObject("headers", new JSObject());
                for (Iterator<String> keys = headers.keys(); keys.hasNext(); ) {
                    String key = keys.next();
                    connection.setRequestProperty(key, headers.getString(key));
                }
                if (connection.getResponseCode() != 200) throw new Exception("HTTP " + connection.getResponseCode());
                long total = connection.getContentLengthLong(), done = 0, reportedAt = 0;
                target.getParentFile().mkdirs();
                try (InputStream in = connection.getInputStream(); FileOutputStream out = new FileOutputStream(part)) {
                    byte[] buffer = new byte[64 * 1024];
                    int read;
                    while ((read = in.read(buffer)) != -1) {
                        out.write(buffer, 0, read);
                        done += read;
                        // At most ten events a second keeps the bridge quiet on fast links.
                        long now = System.currentTimeMillis();
                        if (now - reportedAt >= 100 || done == total) { reportedAt = now; progress(id, done, total); }
                    }
                }
                if (!part.renameTo(target)) throw new Exception("rename failed");
                JSObject ret = new JSObject();
                ret.put("path", target.getPath());
                call.resolve(ret);
            } catch (Exception e) {
                //noinspection ResultOfMethodCallIgnored
                part.delete();
                call.reject(connections.containsKey(id) ? (e.getMessage() == null ? "download failed" : e.getMessage()) : "cancelled");
            } finally {
                connections.remove(id);
                if (connection != null) connection.disconnect();
            }
        }).start();
    }

    @PluginMethod
    public void cancel(PluginCall call) {
        HttpURLConnection connection = connections.remove(call.getString("id", ""));
        if (connection != null) new Thread(connection::disconnect).start();
        call.resolve();
    }

    private void progress(String id, long bytes, long total) {
        JSObject event = new JSObject();
        event.put("id", id);
        event.put("bytes", bytes);
        event.put("total", total);
        notifyListeners("progress", event);
    }
}

File: apps/native/ios/App/App/RavenDownloadPlugin.swift
import Foundation
import Capacitor

/// File download with progress and cancel; the Filesystem plugin's downloadFile has neither.
@objc(RavenDownloadPlugin)
public class RavenDownloadPlugin: CAPPlugin, CAPBridgedPlugin, URLSessionDownloadDelegate {
    public let identifier = "RavenDownloadPlugin"
    public let jsName = "RavenDownload"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "download", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cancel", returnType: CAPPluginReturnPromise),
    ]
    private lazy var session = URLSession(configuration: .default, delegate: self, delegateQueue: nil)
    // Touched from plugin calls and the session's delegate queue alike.
    private let lock = NSLock()
    private var tasks: [String: (task: URLSessionDownloadTask, call: CAPPluginCall, target: URL)] = [:]
    private var reportedAt: [String: TimeInterval] = [:]
    private var folder: URL { FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0].appendingPathComponent("media") }

    // download({ id, url, name, headers }) writes Caches/media/<folder>/<file> and resolves { path } when complete.
    // The page names the file but never the location; only http(s) sources are fetched.
    @objc func download(_ call: CAPPluginCall) {
        guard let id = call.getString("id"), let url = URL(string: call.getString("url") ?? ""), let name = call.getString("name"),
              ["http", "https"].contains(url.scheme ?? ""), Self.isSafeName(name) else {
            return call.reject("bad name or url")
        }
        var request = URLRequest(url: url)
        for (key, value) in call.getObject("headers") ?? [:] { request.setValue(value as? String, forHTTPHeaderField: key) }
        let task = session.downloadTask(with: request)
        task.taskDescription = id
        lock.withLock { tasks[id] = (task, call, folder.appendingPathComponent(name)) }
        task.resume()
    }

    // name is <folder>/<file>: one level, no traversal.
    private static func isSafeName(_ name: String) -> Bool {
        let parts = name.split(separator: "/", omittingEmptySubsequences: false)
        return parts.count == 2 && parts.allSatisfy { !$0.isEmpty && $0 != ".." && !$0.hasPrefix(".") }
    }

    @objc func cancel(_ call: CAPPluginCall) {
        lock.withLock { tasks[call.getString("id") ?? ""]?.task.cancel() }
        call.resolve()
    }

    public func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didWriteData bytesWritten: Int64, totalBytesWritten: Int64, totalBytesExpectedToWrite: Int64) {
        // At most ten events a second keeps the bridge quiet on fast links.
        let id = downloadTask.taskDescription ?? "", now = Date().timeIntervalSince1970
        let due = lock.withLock { now - (reportedAt[id] ?? 0) >= 0.1 || totalBytesWritten == totalBytesExpectedToWrite }
        guard due else { return }
        lock.withLock { reportedAt[id] = now }
        notifyListeners("progress", data: ["id": id, "bytes": totalBytesWritten, "total": totalBytesExpectedToWrite])
    }

    public func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
        guard let id = downloadTask.taskDescription, let entry = lock.withLock({ tasks.removeValue(forKey: id) }) else { return }
        let status = (downloadTask.response as? HTTPURLResponse)?.statusCode ?? 0
        guard status == 200 else { return entry.call.reject("HTTP \(status)") }
        do {
            try FileManager.default.createDirectory(at: entry.target.deletingLastPathComponent(), withIntermediateDirectories: true)
            try? FileManager.default.removeItem(at: entry.target)
            try FileManager.default.moveItem(at: location, to: entry.target)
            entry.call.resolve(["path": entry.target.path])
        } catch {
            entry.call.reject(error.localizedDescription)
        }
    }

    public func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        guard let id = task.taskDescription else { return }
        let entry = lock.withLock { () -> (task: URLSessionDownloadTask, call: CAPPluginCall, target: URL)? in
            reportedAt.removeValue(forKey: id)
            return tasks.removeValue(forKey: id)
        }
        guard let error = error, let entry = entry else { return }
        entry.call.reject((error as NSError).code == NSURLErrorCancelled ? "cancelled" : error.localizedDescription)
    }
}
